grafast/website/grafserv/servers/nuxt.md

```diff
@@ -40,101 +40,25 @@ import { serv } from "@/server/grafserv/serv";
 export default eventHandler((event) => serv.handleEventStreamEvent(event));
 ```
 
-# Experimental
-
-## Websocket support
-
-Nitro and h3 does not yet support WebSocket.
-
-An unofficial and experimental workaround consists to create a nuxt module:
-
-```ts title="modules/grafserv/index.ts"
-// nuxt auto-register modules located in `modules/*.ts` or `modules/*/index.ts`
-
-import { defineNuxtModule, addServerPlugin } from "@nuxt/kit";
-
-import httpProxy from "http-proxy";
-
-export default defineNuxtModule({
-  async setup(options, nuxt) {
-    /**
-     * Register websockets in DEVELOPMENT mode.
-     */
-    if (nuxt.options.dev) {
-      // hook called in development only
-      nuxt.hook("listen", (devServer) => {
-        // create a proxy for routing ws to runtime server created in dev plugin
-        const proxy = httpProxy.createProxy({
-          target: {
-            host: "localhost",
-            port: 3100,
-          },
-        });
-        // registering ws on devServer
-        devServer.on("upgrade", (req, socket, head) => {
-          // routing ws by path
-          switch (req.url) {
-            case "/api/graphql":
-              // proxy websocket to runtime server created in dev plugin
-              proxy.ws(req, socket, head);
-              break;
-            default:
-              socket.destroy();
-          }
-        });
-      });
-      // Registering runtime plugin for dev
-      addServerPlugin("@/modules/grafserv/ws-dev");
-    }
-
-    /**
-     * Register websockets in PRODUCTION mode.
-     */
-    if (!nuxt.options.dev)
-      // Registering runtime plugin for production
-      addServerPlugin("@/modules/grafserv/ws");
-  },
-});
-```
-
-and two Nitro plugins (one for dev, and one for prod)
+## Websocket support (need h3@^1.13.0)
 
-```ts title="modules/grafserv/ws-dev.ts"
-import { Server } from "http";
+```ts title="server/api/graphql-ws.ts"
+import { defineWebSocketHandler } from "h3";
 import { serv } from "@/server/grafserv/serv";
 
-// plugin running in DEVELOPMENT (runtime)
-export default defineNitroPlugin(async (nitroApp) => {
-  // Create a server for handling websockets
-  const server = new Server().listen({ port: 3100 }, () =>
-    console.log("Runtime server listening on port 3100"),
-  );
-  // Cleanly close server (on leave or HMR before plugin reload)
-  nitroApp.hooks.hookOnce("close", () => {
-    server.closeAllConnections();
-    server.close((err) =>
-      err
-        ? console.warn("Runtime server wrongly closed", err)
-        : console.log("Runtime server stopped"),
-    );
-  });
-  // Attaching ws to server
-  serv.attachWebsocketsToServer_experimental(server);
-});
+// Create and export the `/api/graphql` websocket handler
+export default defineWebSocketHandler(serv.makeWsHandler());
```

Comment:
Is `defineWebSocketHandler(cb)` effectively the same as `eventHandler({websocket:cb})`?

Reply:
Not quite : `eventHandler({handler:handler, websocket:cb})` is not explicitly documented (maybe not up to date)
see : [https://h3.unjs.io/guide/websocket](https://h3.unjs.io/guide/websocket)

see code on h3 repo:
```ts
/**
 * Define WebSocket event handler.
 *
 * @see https://h3.unjs.io/guide/websocket
 */
export function defineWebSocketHandler(hooks: Partial<WSHooks>) {
  return defineEventHandler({
    handler() {
      throw createError({
        statusCode: 426,
        statusMessage: "Upgrade Required",
      });
    },
    websocket: hooks,
  });
}
```
`defineWebSocketHandler` must be used when only ws on this handler, instead of `eventHandler` (or `defineEventHandler`) when there is at least a standard handler.

Also, it's right to call `defineWebSocketHandler ` in `addTo` because it push this handler **after** the graphql/graphiql handler in the stack, and the _error_ handler will be called only there is no previous handler in the stack (which is not the case because we set graphql/graphiql handler previously).